File: 41-first-missing-positive/first-missing-positive.cpp
class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        int n = nums.size();
        for(int i=0; i<n; i+=1){
            int num = nums[i];

            while (num>=1 && num<=n && nums[num-1]!=num && num!=i+1){
                swap(nums[i], nums[num-1]);
                num = nums[i];
            }
        }

        for (int i=0; i<n; i+=1){
            if (nums[i] != i+1){
                return i+1;
            }
        }

        return n+1;
    }
};

File: 41-first-missing-positive/first-missing-positive.go
func firstMissingPositive(nums []int) int {
    n := len(nums)
    for i := range nums {
        if nums[i] <= 0 || nums[i] > n {
            nums[i] = n+1
        }
    }

    for i := 0; i < n; i++ {
        num := abs(nums[i])
        if num > n {
            continue
        }
        if nums[num-1] > 0 {
            nums[num-1] *= -1
        }
    }

    for i := 0; i < n; i++ {
        if nums[i] >= 0 {
            return i + 1
        }
    }

    return n + 1
}

func abs(x int) int {
    if x < 0 {
        return -x
    }
    return x
}